src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/SourceControl/VsSourceControlTracker.cs

```diff
@@ -111,20 +110,30 @@ private void StartTracking()
                 return;
             }
 
-            uint cookie;
-            ProjectTracker.AdviseTrackProjectDocumentsEvents(_projectDocumentListener, out cookie);
-            _trackingCookie = cookie;
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                uint cookie;
+                ProjectTracker.AdviseTrackProjectDocumentsEvents(_projectDocumentListener, out cookie);
+                _trackingCookie = cookie;
+            });
         }
 
-        [SuppressMessage("Microsoft.VisualStudio.Threading.Analyzers", "VSTHRD010", Justification = "NuGet/Home#4833 Baseline")]
         private void StopTracking()
         {
             if (!IsTracking)
             {
                 return;
             }
 
-            ProjectTracker.UnadviseTrackProjectDocumentsEvents((uint)_trackingCookie);
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                ProjectTracker.UnadviseTrackProjectDocumentsEvents((uint)_trackingCookie);
```

Comment:
>UnadviseTrackProjectDocumentsEvents [](start = 31, length = 35)

off topic: `ErrorHandler.ThrowOnFailure` to avoid hiding bugs in your Unadvise logic.

Reply:
Also you could use `.Value` instead of a cast.

---
In reply to: [192170484](https://github.com/NuGet/NuGet.Client/pull/2269#discussion_r192170484) [](ancestors = 192170484)